hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java

```diff
@@ -264,54 +288,55 @@ private boolean preHandleStillValid(@Nullable final Object metadata) {
             @NonNull final HederaState state,
             @NonNull final PreHandleResult previousResult,
             @NonNull final Configuration configuration)
             throws PreCheckException {
-        final var txBody = previousResult.txInfo().txBody();
+        final var txInfo = previousResult.txInfo();
+        final var txBody = txInfo.txBody();
+        final var sigPairs = txInfo.signatureMap().sigPairOrElse(emptyList());
+        final var signedBytes = txInfo.signedBytes();
 
         // extract keys and hollow accounts again
         final var storeFactory = new ReadableStoreFactory(state);
         final var context = new PreHandleContextImpl(storeFactory, txBody, configuration);
         dispatcher.dispatchPreHandle(context);
 
-        // setup result with payer key (which has always been verified during preHandle)
-        final var previousVerifications = previousResult.verificationResults();
-        final var newVerifications = new HashMap<Key, SignatureVerificationFuture>();
-        newVerifications.put(previousResult.payerKey(), previousVerifications.get(previousResult.payerKey()));
+        // prepare signature verification
+        final var verifications = new HashMap<Key, SignatureVerificationFuture>();
+        final var payerKey = previousResult.payerKey();
+        verifications.put(payerKey, previousResult.verificationResults().get(payerKey));
 
-        // add non-payer keys
-        final var originals = previousResult.txInfo().signatureMap().sigPairOrElse(emptyList());
+        // expand all keys
         final var expanded = new HashSet<ExpandedSignaturePair>();
-        final var nonPayerKeys = context.requiredNonPayerKeys();
-        for (final var key : nonPayerKeys) {
-            final var found = previousVerifications.get(key);
-            if (found != null) {
-                newVerifications.put(key, found);
-            } else {
-                signatureExpander.expand(key, originals, expanded);
+        signatureExpander.expand(context.requiredNonPayerKeys(), sigPairs, expanded);
+        signatureExpander.expand(context.optionalNonPayerKeys(), sigPairs, expanded);
+
+        // remove all keys that were already verified
+        for (final var it = expanded.iterator(); it.hasNext(); ) {
+            final var entry = it.next();
+            final var oldVerification = previousResult.verificationResults().get(entry.key());
+            if (oldVerification != null) {
+                verifications.put(oldVerification.key(), oldVerification);
+                it.remove();
             }
         }
 
-        // start verification of any key that was not found in the previous result
+        // start verification for remaining keys
         if (!expanded.isEmpty()) {
-            newVerifications.putAll(
-                    signatureVerifier.verify(previousResult.txInfo().signedBytes(), expanded));
+            verifications.putAll(signatureVerifier.verify(signedBytes, expanded));
```

Comment:
This is a bit confusing to me.
How does this work with things like Threshold keys?  If I understand the logic (and I probably do not), then we verify all the expanded (i.e. primitive) keys, and put them all in the verifications, but where do we check that a threshold is met, or that a parent key has enough valid child keys?
A scenario
>Assume File service adds a single key to required keys, and that is a keylist[3] key with one child that is a threshold[5,15] key and 6 of those are keylist[2..12] keys.
>We expand the keys, and get 73 total primitive keys.
>> 36 of those keys validate correctly.
>
> How do we know if that set of 36 is sufficient to make the top-level "required" key valid?
>> Is that handled by including the top key in the expanded set, or is there some other code that ensures the thresholds and key lists are satisfied?

Reply:
This code segment only initiates the verification of all primitive keys in background threads. The actual check is done elsewhere (lines 153-162).